src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java

```diff
@@ -0,0 +1,91 @@
+import org.jabref.logic.importer.util.GrobidServiceException;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.util.FileUpdateMonitor;
+import org.jabref.preferences.JabRefPreferences;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GrobidCitationFetcher implements SearchBasedFetcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GrobidCitationFetcher.class);
+    private ImportFormatPreferences importFormatPreferences;
+    private FileUpdateMonitor fileUpdateMonitor;
+    private GrobidService grobidService;
+    private List<String> failedEntries = new ArrayList<>();
+
+    public GrobidCitationFetcher(ImportFormatPreferences importFormatPreferences, FileUpdateMonitor fileUpdateMonitor, JabRefPreferences jabRefPreferences) {
+        this.importFormatPreferences = importFormatPreferences;
+        this.fileUpdateMonitor = fileUpdateMonitor;
+        grobidService = new GrobidService(jabRefPreferences);
+    }
+
+    /**
+     * Passes request to grobid server, using consolidateCitations option to improve result.
+     * Takes a while, since the server has to look up the entry.
+     */
+    private String parseUsingGrobid(String plainText) throws FetcherException {
+        try {
+            return grobidService.processCitation(plainText, 1);
+        } catch (GrobidServiceException e) {
+            throw new FetcherException("The Pipeline failed to get the results from the GROBID client", e);
+        }
+    }
+
+    private Optional<BibEntry> parseBibToBibEntry(String bibtexString) throws FetcherException {
+        try {
+            return BibtexParser.singleFromString(bibtexString,
+                    importFormatPreferences, fileUpdateMonitor);
+        } catch (ParseException e) {
+            throw new FetcherException("Jabref failed to extract a BibEntry form bibtexString.", e);
+        }
+    }
+
+    @Override
+    public List<BibEntry> performSearch(String query) throws FetcherException {
+        TreeSet<String> plainReferences = new TreeSet<>();
+        String[] plainReferencesArray = query.split(";;");
+        for (int i = 0; i < plainReferencesArray.length; i++) {
+            plainReferences.add(plainReferencesArray[i].trim());
+        }
+        plainReferences.remove("");
+        if (plainReferences.size() == 0) {
+            throw new FetcherException("Your entered References are empty.");
+        } else {
+            ArrayList<BibEntry> resultsList = new ArrayList<>();
+            for (String reference: plainReferences) {
+                parseBibToBibEntry(parseUsingGrobid(reference)).ifPresentOrElse(resultsList::add, () -> failedEntries.add(reference));
+            }
+            return resultsList;
+        }
+    }
+
+    @Override
+    public String getName() {
+        return "GROBID";
+    }
+
+  /**
+   * This method is only used for testing purposes.
+   */
+  public void setGrobidService(GrobidService grobidService) {
```

Comment:
If you need to add methods to make your code testable, then this is usually a good moment to refactor your code. Why should your test be allowed to do something that nobody else is allowed to? In this case, it's better to add the `grobidService` as a constructor argument (with a default constructor falling back to `new GrobidService(...)`)

Reply:
58eb443a1b60784567ec50d8454d6b2d0bc6b876